src/System.Net.Http/src/uap/System/Net/HttpClientHandler.cs

```diff
@@ -684,6 +720,81 @@ private static bool InitRTNoCacheSupported()
+                "ServerCustomValidationRequested");
+        }
+
+        private void RTServerCertificateCallback(RTHttpBaseProtocolFilter sender, RTHttpServerCustomValidationRequestedEventArgs args)
+        {
+            // Convert WinRT certificate to .NET certificate.
+            X509Certificate2 serverCert = ConvertPublicKeyCertificate(args.ServerCertificate);
+
+            // Create .NET X509Chain from the WinRT information. We need to rebuild the chain since WinRT only
+            // gives us an array of intermediate certificates and not a X509Chain object.
+            var serverChain = new X509Chain();
+            SslPolicyErrors sslPolicyErrors = SslPolicyErrors.None;
+            foreach (RTCertificate cert in args.ServerIntermediateCertificates)
+            {
+                serverChain.ChainPolicy.ExtraStore.Add(ConvertPublicKeyCertificate(cert));
+            }
+            serverChain.ChainPolicy.RevocationMode = X509RevocationMode.Online; // WinRT always checks revocation.
+            serverChain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+            // Authenticate the remote party: (e.g. when operating in client mode, authenticate the server).
+            serverChain.ChainPolicy.ApplicationPolicy.Add(s_serverAuthOid);
+            if (!serverChain.Build(serverCert))
+            {
+                sslPolicyErrors |= SslPolicyErrors.RemoteCertificateChainErrors;
+            }
+
+            // Determine name-mismatch error from the existing WinRT information since .NET X509Chain.Build does not
+            // return that in the X509Chain.ChainStatus fields.
+            foreach (RTChainValidationResult result in args.ServerCertificateErrors)
+            {
+                if (result == RTChainValidationResult.InvalidName)
+                {
+                    sslPolicyErrors |= SslPolicyErrors.RemoteCertificateNameMismatch;
+                    break;
+                }
+            }
+
+            // Get the .NET HttpRequestMessage we saved in the property bag of the WinRT HttpRequestMessage.
+            HttpRequestMessage request = (HttpRequestMessage)args.RequestMessage.Properties[RequestMessageLookupKey];
+
+            // Call the .NET callback.
+            bool success = false;
+            try
+            {
+                success = _serverCertificateCustomValidationCallback(request, serverCert, serverChain, sslPolicyErrors);
+            }
+            catch (Exception ex)
+            {
+                // Save the exception info. We will return it later via the SendAsync response processing.
+                args.RequestMessage.Properties.Add(
+                    SavedExceptionDispatchInfoLookupKey,
+                    ExceptionDispatchInfo.Capture(ex));
+            }
+            finally
+            {
+                serverChain.Dispose();
+                serverCert.Dispose();
+            }
+
+            if (!success)
+            {
+                args.Reject();
```

Comment:
I don't like that the API requires call to Reject instead of one to Accept.  Not yours, I know.

But to make sure that no early returns get added to this function to make it dangerous I think it would be useful to make the work be done in a helper method, then this is a simple

```C#
if (!HandleCustomCallback(args))
{
    args.Reject();
}
```

(It'd possibly be better if each piece that needed data from args was fed the data without args leaving this function, so it's clear that no one else alters it)